sklearn/experimental/tests/test_column_transformer.py

```diff
@@ -0,0 +1,303 @@
+    X_res_both = X_array
+
+    # scalar
+    ct = ColumnTransformer([('trans', Trans(), 0)])
+    assert_array_equal(ct.fit_transform(X_array), X_res_first1D)
+    assert_array_equal(ct.fit(X_array).transform(X_array), X_res_first1D)
+
+    ct = ColumnTransformer([('trans', Trans(), [0])])
+    assert_array_equal(ct.fit_transform(X_array), X_res_first2D)
+    assert_array_equal(ct.fit(X_array).transform(X_array), X_res_first2D)
+
+    # list
+    ct = ColumnTransformer([('trans', Trans(), [0, 1])])
+    assert_array_equal(ct.fit_transform(X_array), X_res_both)
+    assert_array_equal(ct.fit(X_array).transform(X_array), X_res_both)
+
+    ct = ColumnTransformer([('trans1', Trans(), [0]),
+                            ('trans2', Trans(), [1])])
+    assert_array_equal(ct.fit_transform(X_array), X_res_both)
+    assert_array_equal(ct.fit(X_array).transform(X_array), X_res_both)
+
+    # slice
+    ct = ColumnTransformer([('trans', Trans(), slice(0, 1))])
+    assert_array_equal(ct.fit_transform(X_array), X_res_first2D)
+    assert_array_equal(ct.fit(X_array).transform(X_array), X_res_first2D)
+
+    ct = ColumnTransformer([('trans', Trans(), slice(0, 2))])
+    assert_array_equal(ct.fit_transform(X_array), X_res_both)
+    assert_array_equal(ct.fit(X_array).transform(X_array), X_res_both)
+
+    # boolean mask
+    ct = ColumnTransformer([('trans', Trans(), np.array([True, False]))])
+    assert_array_equal(ct.fit_transform(X_array), X_res_first2D)
+    assert_array_equal(ct.fit(X_array).transform(X_array), X_res_first2D)
+
+    # test with transformer_weights
+    transformer_weights = {'trans1': .1, 'trans2': 10}
+    both = ColumnTransformer([('trans1', Trans(), [0]),
+                              ('trans2', Trans(), [1])],
+                             transformer_weights=transformer_weights)
+    res = np.vstack([transformer_weights['trans1'] * X_res_first1D,
+                     transformer_weights['trans2'] * X_res_second1D]).T
+    assert_array_equal(both.fit_transform(X_array), res)
+    assert_array_equal(both.fit(X_array).transform(X_array), res)
+
+    both = ColumnTransformer([('trans', Trans(), [0, 1])],
+                             transformer_weights={'trans': .1})
+    assert_array_equal(both.fit_transform(X_array), 0.1 * X_res_both)
+    assert_array_equal(both.fit(X_array).transform(X_array), 0.1 * X_res_both)
+
+
+def test_column_transformer_dataframe():
+    try:
+        import pandas as pd
+    except ImportError:
+        raise SkipTest("pandas is not installed: skipping ColumnTransformer "
+                       "tests for DataFrames.")
+
+    X_array = np.array([[0, 1, 2], [2, 4, 6]]).T
+    X_df = pd.DataFrame(X_array, columns=['first', 'second'])
+
+    X_res_first1D = np.array([0, 1, 2])
+    X_res_first2D = X_res_first1D.reshape(-1, 1)
+    X_res_both = X_array
+
+    # String keys: label based
+
+    # scalar
+    ct = ColumnTransformer([('trans', Trans(), 'first')])
```

Comment:
Insofar as is possible can the tests be refactored so that we are essentially checking the equivalence between different indexing schemes / data types?